IDE support for Dotty via the Language Server Protocol, including a Visual Studio Code extension

smarter opened the pull request:
This PR introduces IDE support to Dotty, this requires several components (see each individual commit messages for more detailed explanations):
- A set of experimental APIs in the compiler designed for IDE usage: `dotty.tools.dotc.interactive.*`. These APIs were developed for the Dotty Language Server but could evolve to support other usages like the REPL or other IDEs.
- An implementation of the [Language Server Protocol](https://github.com/Microsoft/language-server-protocol) for Dotty, this allow any editor supporting the LSP to get great Dotty support with a minimal amount of work (see http://langserver.org/ for a list of LSP implementations).
- A [Visual Studio Code](http://code.visualstudio.com/) extension to run the Dotty Language Server.
- New commands in the sbt-dotty plugin to setup everything needed to run the Dotty Language Server and start Visual Studio Code.

# Quick start

1. Install [Visual Studio Code](http://code.visualstudio.com/)
2. In your own project compiled with dotty, make sure sbt-dotty 0.1.0-RC5 is used, and run:
```shell
sbt launchIDE
```

# Status

## Fully supported features:
- Typechecking as you type to show compiler errors/warnings
- Type information on hover
- Go to definition (in the current project)
- Find all references

## Partially working features:
- Completion
- Renaming
- Go to definition in external projects

## Unimplemented features:
- Documentation on hover
- Formatting code (requires integrating with scalafmt)
- Quick fixes (probably by integrating with scalafix)

## Current limitations, to be fixed:
- Projects should be compiled with sbt before starting the IDE, this is
  automatically done for you if you run `sbt launchIDE`.
- Once the IDE is started, source files that are not opened in the IDE
  should not be modified in some other editor, the IDE won't pick up
  these changes.
- Not all compiler errors/warnings are displayed, just those occuring
  during typechecking.

## Future
There is much more that could be done, in particular see https://github.com/lampepfl/dotty/issues/1523 for a lot of potential improvements that we should keep in mind.

smarter commented:
/cc @fommil @dragos @sschaef

fommil commented:
nice! btw, we're going to add language server to ensime. This will serve as an excellent reference implementation for ensime 3.0. Hmm, I guess scastie might need to use this for dotty support.

felixmulder commented on docs/docs/usage/ide-support.md:
```diff
@@ -0,0 +1,58 @@
+Dotty comes built-in with the Dotty Language Server, an implementation of the
+[Language Server Protocol](https://github.com/Microsoft/language-server-protocol),
+which means that any editor that implements support for the LSP can get a great
+Dotty integration. Currently, the only IDE we officially support
```
"can get a great Dotty integration." => "can work as a Dotty IDE; to the extent of the protocol."

felixmulder commented on docs/docs/usage/ide-support.md:
```diff
@@ -0,0 +1,58 @@
+
+## Current limitations, to be fixed:
+- Projects should be compiled with sbt before starting the IDE, this is
+  automatically done for you if you run `sbt launchIDE`.
```
Does it cancel if the project does not compile?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+
+  private[this] var myDrivers: mutable.Map[ProjectConfig, InteractiveDriver] = _
+
+  def drivers: mutable.Map[ProjectConfig, InteractiveDriver] = thisServer.synchronized {
```
Why is this mutable?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+
+  def drivers: mutable.Map[ProjectConfig, InteractiveDriver] = thisServer.synchronized {
+    if (myDrivers == null) {
+      assert(rootUri != null, "`drivers` cannot be called before `initialize`")
```
Can't we get it to call `initialize` for us if it wasn't done before?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+    c.setReferencesProvider(true)
+    c.setCompletionProvider(new CompletionOptions(
+      /* resolveProvider = */ false,
+      /* triggerCharacters = */ List(".").asJava))
```
I thought Java was big on the builder pattern  :joy:

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+    val driver = driverFor(uri)
+
+    val change = params.getContentChanges.get(0)
+    assert(change.getRange == null, "TextDocumentSyncKind.Incremental support is not implemented")
```
Why is this an assert? Is it to show an incorrectly set up client an error? There isn't something like "client error" that can be sent back instead?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+  }
+
+  override def didChangeConfiguration(params: DidChangeConfigurationParams): Unit =
+    /*thisServer.synchronized*/ {}
```
Maybe a docstring here?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+    /*thisServer.synchronized*/ {}
+
+  override def didChangeWatchedFiles(params: DidChangeWatchedFilesParams): Unit =
+    /*thisServer.synchronized*/ {}
```
?

felixmulder commented on language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala:
```diff
@@ -0,0 +1,438 @@
+    /*thisServer.synchronized*/ {}
+
+  override def didSave(params: DidSaveTextDocumentParams): Unit =
+    /*thisServer.synchronized*/ {}
```
?

felixmulder commented on language-server/src/dotty/tools/languageserver/Main.scala:
```diff
@@ -0,0 +1,75 @@
+/** Run the Dotty Language Server.
+ *
+ * This is designed to be started from an editor supporting the Language Server
+ * Protocol, the easiest way to fetch and run this is to use `coursier`:
```
Alignment

felixmulder commented on language-server/src/dotty/tools/languageserver/Main.scala:
```diff
@@ -0,0 +1,75 @@
+ * This is designed to be started from an editor supporting the Language Server
+ * Protocol, the easiest way to fetch and run this is to use `coursier`:
+ *
+ *   coursier launch $artifact -M dotty.tools.languageserver.Main -- -stdio
```
` ```shell`?

smarter replied:
I tried it, and it doesn't do any highlighting on this line, so no point in using this syntax.

felixmulder commented on compiler/src/dotty/tools/dotc/interactive/Interactive.scala:
```diff
@@ -0,0 +1,183 @@
+   */
+  @tailrec def sourceSymbol(sym: Symbol)(implicit ctx: Context): Symbol =
+    if (!sym.exists)
+      sym
```
I'll leave that up to you.

odersky replied:
I think either works fine.

odersky commented:
For what it's worth, I could not launch it. I dowloaded VScode and start that one OK. But then I got
this:

    java.io.IOException: Cannot run program "code": error=2, No such file or directory

Full transcript:

/Users/odersky/workspace/dotty> sbt
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
[info] Loading project definition from /Users/odersky/workspace/dotty/project/project
[info] Updating {file:/Users/odersky/workspace/dotty/project/project/}dotty-build-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Compiling 1 Scala source to /Users/odersky/workspace/dotty/project/project/target/scala-2.10/sbt-0.13/classes...
[info] Loading project definition from /Users/odersky/workspace/dotty/project
[info] Updating {file:/Users/odersky/workspace/dotty/project/}dotty-build...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] downloading https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-databind/2.8.8/jackson-databind-2.8.8.jar ...
[info] 	[SUCCESSFUL ] com.fasterxml.jackson.core#jackson-databind;2.8.8!jackson-databind.jar(bundle) (918ms)
[info] downloading https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-annotations/2.8.0/jackson-annotations-2.8.0.jar ...
[info] 	[SUCCESSFUL ] com.fasterxml.jackson.core#jackson-annotations;2.8.0!jackson-annotations.jar(bundle) (148ms)
[info] downloading https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.8.8/jackson-core-2.8.8.jar ...
[info] 	[SUCCESSFUL ] com.fasterxml.jackson.core#jackson-core;2.8.8!jackson-core.jar(bundle) (252ms)
[info] Done updating.
[info] Compiling 5 Scala sources and 1 Java source to /Users/odersky/workspace/dotty/project/target/scala-2.10/sbt-0.13/classes...
[warn] /Users/odersky/workspace/dotty/project/Build.scala:201: method apply in object ScalaInstance is deprecated: Does not handle modularized Scala.  Use a variant that only accepts all jars.
[warn]       ScalaInstance(scalaVersion.value, libraryJar, compilerJar, otherDependencies: _*)(state.value.classLoaderCache.apply)
[warn]       ^
[warn] one warning found
[info] Resolving key references (11645 settings) ...
[info] Set current project to dotty (in build file:/Users/odersky/workspace/dotty/)
> launchIDE
[info] Updating {file:/Users/odersky/workspace/dotty/}scalap...
[info] Updating {file:/Users/odersky/workspace/dotty/}scala-compiler...
[info] Updating {file:/Users/odersky/workspace/dotty/}scala-reflect...
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-library-bootstrapped...
[info] Resolving org.scala-lang#scala-library;2.11.11 ...
[info] Done updating.
[info] Resolving org.scala-sbt#test-interface;1.0 ...
[info] Packaging /Users/odersky/workspace/dotty/sbt-bridge/target/dotty-sbt-bridge-0.1.1-bin-SNAPSHOT-nonbootstrapped-sources.jar ...
[info] Done updating.
[info] Done packaging.
[info] Updating {file:/Users/odersky/workspace/dotty/}scala-library...
[info] Done updating.
[info] Resolving org.scala-lang.modules#scala-parser-combinators_2.11;1.0.4 ...
[info] Done updating.
[info] Resolving org.scala-lang#scala-library;2.11.11 ...
[info] Done updating.
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-compiler-bootstrapped...
[info] Resolving org.scala-sbt#test-interface;1.0 ...
[info] Done updating.
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-language-server...
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-bootstrapped...
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-sbt-bridge-bootstrapped...
[info] Updating {file:/Users/odersky/workspace/dotty/}dotty-doc-bootstrapped...
[info] Resolving org.scala-lang.modules#scala-xml_2.11;1.0.1 ...
[info] Done updating.
[info] Resolving org.eclipse.lsp4j#org.eclipse.lsp4j;0.2.0.M7 ...
[info] Packaging /Users/odersky/workspace/dotty/library/target/scala-2.11/dotty-library_2.11-0.1.1-bin-SNAPSHOT-nonbootstrapped.jar ...
[info] Done packaging.
[info] Resolving org.eclipse.lsp4j#org.eclipse.lsp4j.jsonrpc;0.2.0.M7 ...
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[info] Resolving com.google.code.gson#gson-parent;2.7 ...
[info] Packaging /Users/odersky/workspace/dotty/compiler/target/scala-2.11/dotty-compiler_2.11-0.1.1-bin-SNAPSHOT-nonbootstrapped.jar ...
[info] Resolving com.google.guava#guava;[14.0,19) ...
[info] Done packaging.
[info] Compiling 26 Scala sources and 136 Java sources to /Users/odersky/workspace/dotty/library/../out/bootstrap/dotty-library-bootstrapped/scala-0.1/classes...
[info] Resolving com.fasterxml.jackson.core#jackson-core;2.8.8 ...
[info] downloading https://repo1.maven.org/maven2/org/eclipse/lsp4j/org.eclipse.lsp4j/0.2.0.M7/org.eclipse.lsp4j-0.2.0.M7.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.lsp4j#org.eclipse.lsp4j;0.2.0.M7!org.eclipse.lsp4j.jar (468ms)
[info] downloading https://repo1.maven.org/maven2/org/eclipse/lsp4j/org.eclipse.lsp4j.generator/0.2.0.M7/org.eclipse.lsp4j.generator-0.2.0.M7.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.lsp4j#org.eclipse.lsp4j.generator;0.2.0.M7!org.eclipse.lsp4j.generator.jar (131ms)
[info] downloading https://repo1.maven.org/maven2/org/eclipse/lsp4j/org.eclipse.lsp4j.jsonrpc/0.2.0.M7/org.eclipse.lsp4j.jsonrpc-0.2.0.M7.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.lsp4j#org.eclipse.lsp4j.jsonrpc;0.2.0.M7!org.eclipse.lsp4j.jsonrpc.jar (215ms)
[info] downloading https://repo1.maven.org/maven2/org/eclipse/xtend/org.eclipse.xtend.lib/2.12.0/org.eclipse.xtend.lib-2.12.0.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.xtend#org.eclipse.xtend.lib;2.12.0!org.eclipse.xtend.lib.jar (321ms)
[info] downloading https://repo1.maven.org/maven2/com/google/code/gson/gson/2.7/gson-2.7.jar ...
[info] 	[SUCCESSFUL ] com.google.code.gson#gson;2.7!gson.jar (339ms)
[info] downloading https://repo1.maven.org/maven2/org/eclipse/xtext/org.eclipse.xtext.xbase.lib/2.12.0/org.eclipse.xtext.xbase.lib-2.12.0.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.xtext#org.eclipse.xtext.xbase.lib;2.12.0!org.eclipse.xtext.xbase.lib.jar (238ms)
[info] downloading https://repo1.maven.org/maven2/org/eclipse/xtend/org.eclipse.xtend.lib.macro/2.12.0/org.eclipse.xtend.lib.macro-2.12.0.jar ...
[info] 	[SUCCESSFUL ] org.eclipse.xtend#org.eclipse.xtend.lib.macro;2.12.0!org.eclipse.xtend.lib.macro.jar (174ms)
[info] downloading https://repo1.maven.org/maven2/com/google/guava/guava/18.0/guava-18.0.jar ...
[info] 	[SUCCESSFUL ] com.google.guava#guava;18.0!guava.jar(bundle) (1797ms)
[info] Done updating.
[info] Resolving org.scala-sbt#test-interface;1.0 ...
[info] downloading https://repo1.maven.org/maven2/com/fasterxml/jackson/dataformat/jackson-dataformat-yaml/2.8.8/jackson-dataformat-yaml-2.8.8.jar ...
[info] 	[SUCCESSFUL ] com.fasterxml.jackson.dataformat#jackson-dataformat-yaml;2.8.8!jackson-dataformat-yaml.jar(bundle) (161ms)
[info] Done updating.
[info] Resolving org.scalaz#scalaz-effect_2.11;7.0.6 ...
[info] Done updating.
[info] 'dotty-sbt-bridge' not yet compiled for Scala 0.1.1-bin-SNAPSHOT-nonbootstrapped. Compiling...
[info]   Compilation completed in 3.384 s
[info] Compiling 316 Scala sources and 1 Java source to /Users/odersky/workspace/dotty/compiler/../out/bootstrap/dotty-compiler-bootstrapped/scala-0.1/classes...
sig of dotty.tools.dotc.core.NameOps.NameDecorator[N](
  this.functionName.++(m.paramInfos.length.toString())
).specializedFor: Signature(List(scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, dotty.tools.dotc.core.Contexts$.Context),dotty.tools.dotc.core.Names$.Name)
sig of dotty.tools.dotc.core.NameOps.NameDecorator[
  dotty.tools.dotc.core.Names.TermName^
](this.functionName.++(m.paramInfos.length.toString())).specializedFor: Signature(List(scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, dotty.tools.dotc.core.Contexts$.Context),dotty.tools.dotc.core.Names$.TermName)
[warn] -- Deprecation Warning: /Users/odersky/workspace/dotty/compiler/../out/bootstrap/dotty-compiler-bootstrapped/scala-0.1/src_managed/main/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:1394:12 
[warn] 1394 |        new asm.Handle(invokeStyle,
[warn]      |            ^^^^^^^^^^
[warn]      |constructor Handle in class Handle is deprecated: see corresponding Javadoc for more information.
[warn] -- Deprecation Warning: /Users/odersky/workspace/dotty/compiler/../out/bootstrap/dotty-compiler-bootstrapped/scala-0.1/src_managed/main/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:1425:8 
[warn] 1425 |    new asm.Handle(asm.Opcodes.H_INVOKESTATIC,
[warn]      |        ^^^^^^^^^^
[warn]      |constructor Handle in class Handle is deprecated: see corresponding Javadoc for more information.
[warn] -- Deprecation Warning: /Users/odersky/workspace/dotty/compiler/../out/bootstrap/dotty-compiler-bootstrapped/scala-0.1/src_managed/main/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:414:14 
[warn] 414 |      jmethod.visitMethodInsn(Opcodes.INVOKEDYNAMIC, owner, name, desc)
[warn]     |      ^^^^^^^^^^^^^^^^^^^^^^^
[warn]     |method visitMethodInsn in class MethodVisitor is deprecated: see corresponding Javadoc for more information.
sig of dotty.tools.dotc.core.NameOps.NameDecorator[
  dotty.tools.dotc.core.Names.TermName^
](this.functionName.++(m.paramInfos.length.toString())).specializedFor: Signature(List(scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, scala.collection.immutable.List, dotty.tools.dotc.core.Contexts$.Context),dotty.tools.dotc.core.Names$.TermName)
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/config/OutputDirs.scala:110:44 
[warn] 110 |        (outputs filter (isBelow _).tupled) match {
[warn]     |        ^
[warn]     |        match may not be exhaustive.
[warn]     |        
[warn]     |        It would fail on: List(_)
[warn] 111 |          case Nil => Nil
[warn] 112 |          case matches => matches.map(_._1.lookupPathUnchecked(srcPath, false))
[warn] 113 |        }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/repl/ammonite/filters/HistoryFilter.scala:157:25 
[warn] 157 |          filter0.op(ti) match {
[warn]     |          ^
[warn]     |          match may not be exhaustive.
[warn]     |          
[warn]     |          It would fail on: Some(_)
[warn] 158 |            case Some(ts: TermState) =>
[warn] 159 |              prevBuffer = Some(ts.buffer)
[warn] 160 |              Some(ts)
[warn] 161 |            case x => x
[warn] 162 |          }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/transform/Constructors.scala:249:35 
[warn] 249 |    val mappedSuperCalls = vparams match {
[warn]     |                           ^
[warn]     |                      match may not be exhaustive.
[warn]     |                      
[warn]     |                      It would fail on: List(Trees.ValDef(_, _, _), _)
[warn] 250 |      case (outerParam @ ValDef(nme.OUTER, _, _)) :: _ =>
[warn] 251 |        superCalls.map(mapOuter(outerParam.symbol).transform)
[warn] 252 |      case _ => superCalls
[warn] 253 |    }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala:330:12 
[warn] 330 |            case List(IntEqualityTestTreeMaker(_)) => true
[warn]     |            ^
[warn]     |            match may not be exhaustive.
[warn]     |            
[warn]     |            It would fail on: List(_)
[warn] 331 |            case _ => false
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala:362:12 
[warn] 362 |            case List(IntEqualityTestTreeMaker(intValue)) => intValue
[warn]     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[warn]     |            match may not be exhaustive.
[warn]     |            
[warn]     |            It would fail on: List(_)
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala:1209:55 
[warn] 1209 |      val (nonSyntheticCases, defaultOverride) = cases match {
[warn]      |                                                 ^
[warn]      |                                        match may not be exhaustive.
[warn]      |                                        
[warn]      |                                        It would fail on: List(_)
[warn] 1210 |        case init :+ last if isSyntheticDefaultCase(last) => (init, Some(((scrut: Symbol) => last.body)))
[warn] 1211 |        case _                                            => (cases, None)
[warn] 1212 |      }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/typer/Applications.scala:1229:78 
[warn] 1228 |          alts.filter(alt =>
[warn] 1229 |            (alt ne chosen) && resultConforms(alt, pt.resultType)(nestedCtx)) match {
[warn]      |          ^
[warn]      |          match may not be exhaustive.
[warn]      |          
[warn]      |          It would fail on: List(Types.TermRef(_, _), _)
[warn] 1230 |            case Nil => chosen
[warn] 1231 |            case alt2 :: Nil => alt2
[warn] 1232 |            case alts2 =>
[warn] 1233 |              resolveOverloaded(alts2, pt) match {
[warn] 1234 |                case alt2 :: Nil => alt2
[warn] 1235 |                case _ => chosen
[warn] 1236 |              }
[warn] 1237 |          }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/typer/Checking.scala:463:25 
[warn] 463 |        clParamAccessors match {
[warn]     |        ^
[warn]     |        match may not be exhaustive.
[warn]     |        
[warn]     |        It would fail on: List(_)
[warn] 464 |          case List(param) =>
[warn] 465 |            if (param.is(Mutable))
[warn] 466 |              ctx.error("value class parameter must not be a var", param.pos)
[warn] 467 |            if (param.info.isPhantom)
[warn] 468 |              ctx.error("value class parameter must not be phantom", param.pos)
[warn] 469 |          case _ =>
[warn] 470 |            ctx.error("value class needs to have exactly one val parameter", clazz.pos)
[warn] 471 |        }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/typer/RefChecks.scala:69:11 
[warn] 69 |      for ((_, m1 :: m2 :: _) <- (clazz.info member nme.applyDynamic).alternatives groupBy (_.symbol.typeParams.length)) {
[warn]    |           ^
[warn]    |           match may not be exhaustive.
[warn]    |           
[warn]    |           It would fail on: (_, _)
[warn] 70 |        ctx.error("implementation restriction: applyDynamic cannot be overloaded except by methods with different numbers of type parameters, e.g. applyDynamic[T1](method: String)(arg: T1) and applyDynamic[T1, T2](method: String)(arg1: T1, arg2: T2)",
[warn] 71 |          m1.symbol.pos)
[warn] 72 |      }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/typer/RefChecks.scala:512:27 
[warn] 512 |                mismatches match {
[warn]     |                ^
[warn]     |                match may not be exhaustive.
[warn]     |                
[warn]     |                It would fail on: List((_, _), _)
[warn] 513 |                  // Only one mismatched parameter: say something useful.
[warn] 514 |                  case (pa, pc) :: Nil =>
[warn] 515 |                    val abstractSym = pa.typeSymbol
[warn] 516 |                    val concreteSym = pc.typeSymbol
[warn] 517 |                    def subclassMsg(c1: Symbol, c2: Symbol) =
[warn] 518 |                      s": ${c1.showLocated} is a subclass of ${c2.showLocated}, but method parameter types must match exactly."
[warn] 519 |                    val addendum =
[warn] 520 |                      if (abstractSym == concreteSym) {
[warn] 521 |                        val paArgs = pa.argInfos
[warn] 522 |                        val pcArgs = pc.argInfos
[warn] 523 |                        val paConstr = pa.withoutArgs(paArgs)
[warn] 524 |                        val pcConstr = pc.withoutArgs(pcArgs)
[warn] 525 |                        (paConstr, pcConstr) match {
[warn] 526 |                          case (TypeRef(pre1, _), TypeRef(pre2, _)) =>
[warn] 527 |                            if (pre1 =:= pre2) ": their type parameters differ"
[warn] 528 |                            else ": their prefixes (i.e. enclosing instances) differ"
[warn] 529 |                          case _ =>
[warn] 530 |                            ""
[warn] 531 |                        }
[warn] 532 |                      } else if (abstractSym isSubClass concreteSym)
[warn] 533 |                        subclassMsg(abstractSym, concreteSym)
[warn] 534 |                      else if (concreteSym isSubClass abstractSym)
[warn] 535 |                        subclassMsg(concreteSym, abstractSym)
[warn] 536 |                      else ""
[warn] 537 |                    undefined(s"\n(Note that ${pa.show} does not match ${pc.show}$addendum)")
[warn] 538 |                  case xs =>
[warn] 539 |                    undefined(s"\n(The class implements a member with a different type: ${concrete.showDcl})")
[warn] 540 |                }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:186:28 
[warn] 186 |            val what = alts match {
[warn]     |                       ^
[warn]     |                       match may not be exhaustive.
[warn]     |                       
[warn]     |                       It would fail on: List(Symbol(_, _), _)
[warn] 187 |              case Nil =>
[warn] 188 |                name.toString
[warn] 189 |              case sym :: Nil =>
[warn] 190 |                if (sym.owner == pre.typeSymbol) sym.show else sym.showLocated
[warn] 191 |              case _ =>
[warn] 192 |                em"none of the overloaded alternatives named $name"
[warn] 193 |            }
[warn] 14 warnings found
[info] Compiling 47 Scala sources and 3 Java sources to /Users/odersky/workspace/dotty/compiler/../out/bootstrap/dotty-compiler-bootstrapped/scala-0.1/test-classes...
[info] Compiling 45 Scala sources to /Users/odersky/workspace/dotty/doc-tool/../out/bootstrap/dotty-doc-bootstrapped/scala-0.1/classes...
[info] Compiling 2 Scala sources and 1 Java source to /Users/odersky/workspace/dotty/language-server/../out/bootstrap/dotty-language-server/scala-0.1/classes...
[info] Compiling 7 Scala sources to /Users/odersky/workspace/dotty/sbt-bridge/../out/bootstrap/dotty-sbt-bridge-bootstrapped/classes...
[warn] -- Deprecation Warning: /Users/odersky/workspace/dotty/language-server/src/dotty/tools/languageserver/ScalaLanguageServer.scala:131:73 
[warn] 131 |    val rootUri = Option(params.getRootUri).getOrElse("file://" + params.getRootPath)
[warn]     |                                                                  ^^^^^^^^^^^^^^^^^^
[warn]     |method getRootPath in class InitializeParams is deprecated: see corresponding Javadoc for more information.
[warn] one warning found
[warn] -- Deprecation Warning: /Users/odersky/workspace/dotty/compiler/test/dotty/tools/backend/jvm/AsmConverters.scala:232:63 
[warn] 232 |  def unconvertMethodHandle(h: MethodHandle): asm.Handle = new asm.Handle(h.tag, h.owner, h.name, h.desc)
[warn]     |                                                               ^^^^^^^^^^
[warn]     |constructor Handle in class Handle is deprecated: see corresponding Javadoc for more information.
[warn] -- [E029] Match case Unreachable Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/model/comment/CommentParser.scala:96:67 
[warn] 96 |      case SymbolTagRegex(name, sym, body) :: ls if (!inCodeBlock) => {
[warn]     |                                                                   ^
[warn]     |                                                      unreachable code
[warn] 97 |        val key = SymbolTagKey(name, sym)
[warn] 98 |        val value = body :: tags.getOrElse(key, Nil)
[warn] 99 |        parseComment(docBody, tags + (key -> value), Some(key), ls, inCodeBlock)
[warn] 100 |      }
[warn] -- [E029] Match case Unreachable Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/model/comment/CommentParser.scala:102:62 
[warn] 102 |      case SimpleTagRegex(name, body) :: ls if (!inCodeBlock) => {
[warn]     |                                                              ^
[warn]     |                                                      unreachable code
[warn] 103 |        val key = SimpleTagKey(name)
[warn] 104 |        val value = body :: tags.getOrElse(key, Nil)
[warn] 105 |        parseComment(docBody, tags + (key -> value), Some(key), ls, inCodeBlock)
[warn] 106 |      }
[warn] -- [E029] Match case Unreachable Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/model/comment/CommentParser.scala:108:56 
[warn] 108 |      case SingleTagRegex(name) :: ls if (!inCodeBlock) => {
[warn]     |                                                        ^
[warn]     |                                                      unreachable code
[warn] 109 |        val key = SimpleTagKey(name)
[warn] 110 |        val value = "" :: tags.getOrElse(key, Nil)
[warn] 111 |        parseComment(docBody, tags + (key -> value), Some(key), ls, inCodeBlock)
[warn] 112 |      }
[warn] -- [E029] Match case Unreachable Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/model/comment/CommentParser.scala:114:48 
[warn] 114 |      case line :: ls if (lastTagKey.isDefined) => {
[warn]     |                                                ^
[warn]     |                                                unreachable code
[warn] 115 |        val newtags = if (!line.isEmpty) {
[warn] 116 |          val key = lastTagKey.get
[warn] 117 |          val value =
[warn] 118 |            ((tags get key): @unchecked) match {
[warn] 119 |              case Some(b :: bs) => (b + endOfLine + line) :: bs
[warn] 120 |              case None => oops("lastTagKey set when no tag exists for key")
[warn] 121 |            }
[warn] 122 |          tags + (key -> value)
[warn] 123 |        } else tags
[warn] 124 |        parseComment(docBody, newtags, lastTagKey, ls, inCodeBlock)
[warn] 125 |      }
[warn] -- [E029] Match case Unreachable Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/model/comment/CommentParser.scala:127:22 
[warn] 127 |      case line :: ls => {
[warn]     |                      ^
[warn]     |                      unreachable code
[warn] 128 |        if (docBody.length > 0) docBody append endOfLine
[warn] 129 |        docBody append line
[warn] 130 |        parseComment(docBody, tags, lastTagKey, ls, inCodeBlock)
[warn] 131 |      }
[warn] -- [E028] Pattern Match Exhaustivity Warning: /Users/odersky/workspace/dotty/doc-tool/src/dotty/tools/dottydoc/util/MemberLookup.scala:71:21 
[warn] 71 |    (querys, entity) match {
[warn]    |    ^
[warn]    |    match may not be exhaustive.
[warn]    |    
[warn]    |    It would fail on: (Nil, _)
[warn] 72 |      case (xs, NonEntity) => globalLookup
[warn] 73 |      case (x :: Nil, e: Entity with Members) =>
[warn] 74 |        localLookup(e, x)
[warn] 75 |      case (x :: _, e: Entity with Members) if x == entity.name =>
[warn] 76 |        downwardLookup(e, querys)
[warn] 77 |      case (x :: xs, _) =>
[warn] 78 |        if (xs.nonEmpty) globalLookup
[warn] 79 |        else lookup(entity, packages, "scala." + query)
[warn] 80 |    }
[warn] one warning found
[warn] 6 warnings found
[success] Total time: 68 s, completed May 26, 2017 4:47:30 PM
[error] ({.}/*:runCode) java.io.IOException: Cannot run program "code": error=2, No such file or directory
[error] Total time: 0 s, completed May 26, 2017 4:47:30 PM
> exit
/Users/odersky/workspace/dotty> sbt launchIDE
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
[warn] Executing in batch mode.
[warn]   For better performance, hit [ENTER] to switch to interactive mode, or
[warn]   consider launching sbt without any commands, or explicitly passing 'shell'
[info] Loading project definition from /Users/odersky/workspace/dotty/project/project
[info] Loading project definition from /Users/odersky/workspace/dotty/project
[info] Resolving key references (11645 settings) ...
[info] Set current project to dotty (in build file:/Users/odersky/workspace/dotty/)
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[success] Total time: 2 s, completed May 26, 2017 4:48:27 PM
java.io.IOException: Cannot run program "code": error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at dotty.tools.sbtplugin.DottyIDEPlugin$$anonfun$buildSettings$1.apply(DottyIDEPlugin.scala:95)
	at dotty.tools.sbtplugin.DottyIDEPlugin$$anonfun$buildSettings$1.apply(DottyIDEPlugin.scala:92)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:248)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	at dotty.tools.sbtplugin.DottyIDEPlugin$$anonfun$buildSettings$1.apply(DottyIDEPlugin.scala:95)
	at dotty.tools.sbtplugin.DottyIDEPlugin$$anonfun$buildSettings$1.apply(DottyIDEPlugin.scala:92)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] ({.}/*:runCode) java.io.IOException: Cannot run program "code": error=2, No such file or directory
[error] Total time: 0 s, completed May 26, 2017 4:48:27 PM
/Users/odersky/workspace/dotty>

smarter commented:
> java.io.IOException: Cannot run program "code": error=2, No such file or directory

You need `code` (which is the name of the vscode binary) to be in your PATH. Look at where vscode was installed to find it. I'll update the doc to mention that, but unfortunately I don't know what step-by-step instruction to give to do this on every operating system.

felixmulder commented:
```bash
$ brew cask install visual-studio-code
```

Is the quick way on macOS

odersky commented:
@smarter Thanks for the tip. I could launch it now. Looks pretty! I tried it on dotty itself, but so far did not get much to work beyond syntax highlighting. Let's check on Monday together?

smarter commented:
@odersky Sure, did you get any message in the console you launched it from?

odersky commented:
@smarter In the console everything looked normal. Here's what I got:

/Users/odersky/workspace/dotty> sbt launchIDE
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0
[warn] Executing in batch mode.
[warn]   For better performance, hit [ENTER] to switch to interactive mode, or
[warn]   consider launching sbt without any commands, or explicitly passing 'shell'
[info] Loading project definition from /Users/odersky/workspace/dotty/project/project
[info] Loading project definition from /Users/odersky/workspace/dotty/project
[info] Resolving key references (11645 settings) ...
[info] Set current project to dotty (in build file:/Users/odersky/workspace/dotty/)
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[info] Packaging /Users/odersky/workspace/dotty/compiler/target/scala-2.11/dotty-compiler_2.11-0.1.1-bin-SNAPSHOT-nonbootstrapped.jar ...
[info] Done packaging.
[success] Total time: 4 s, completed May 26, 2017 5:17:43 PM
Extension 'lampepfl.dotty' is already installed.
[success] Total time: 0 s, completed May 26, 2017 5:17:43 PM
[warn] Multiple main classes detected.  Run 'show discoveredMainClasses' to see the list
[success] Total time: 1 s, completed May 26, 2017 5:17:44 PM
1. Waiting for source changes... (press enter to interrupt)

smarter commented:
@odersky: Thanks to @liufengyun I fixed a few issues that could have prevented the IDE from running and updated the branch. But the sbt output you're getting makes me wonder if you're on the correct branch, can you check that you're on branch `ide` from staging and not `ide-support` or some other stale branch? Also make sure to get the latest version of the branch after my recent fixes.

smarter commented on compiler/src/dotty/tools/dotc/ast/NavigateAST.scala:
```diff
@@ -74,9 +74,23 @@ object NavigateAST {
+          }))
+
+      if (currentHasPos) {
+        p match {
```
Maybe, but I think it would be even better if we used something higher-level than dealing with productIterator by hand, that would take care of lazy trees, I'll leave a FIXME in the code.

smarter commented on compiler/src/dotty/tools/dotc/ast/NavigateAST.scala:
```diff
@@ -74,9 +74,23 @@ object NavigateAST {
+      val currentHasPos =
+        p.pos.contains(pos) && (!p.pos.isZeroExtent ||
+          (path match {
+            case p1 :: _ => p1.pos.isZeroExtent || !p1.pos.contains(pos)
```
Indeed, I forgot to clean this up, and trying to document it I realized it's broken: I was trying to get the next node in the tree that the path could go through, but this is actually pretty hard to do. I've given up on that and instead added a `skipZeroExtent` parameter in `NavigateAST#pathTo` (that I use in `Interactive#pathTo`): https://github.com/lampepfl/dotty/pull/2532/commits/b3befee6f4b9ce09babfc577c39fdbd6d2b4f299

HerringtonDarkholme commented:
I encounter a NPE when firing up `sbt launcheIDE` in https://github.com/lampepfl/dotty-example-project.

Seems https://github.com/lampepfl/dotty/blob/352c09e481cb4893812c5f49959e587842448fe0/language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala#L83 `drivers.keys.head` throws NPE because drivers is empty. Still investigating what happened here. I will report further info in another issue if I find out why.

smarter commented:
@HerringtonDarkholme could you post the content of the generated .dotty-ide.json file in the project, as well as the output of the `pwd` command in a shell in the project directory? What operating system are you on?

smarter commented:
The exact stacktrace you got would be helpful too

HerringtonDarkholme commented:
I'm on Mac OSX. 

.dotty-ide.json

```
[ {
  "id" : "root/compile",
  "compilerVersion" : "0.1.2-RC1",
  "compilerArguments" : [ ],
  "sourceDirectories" : [ "/test/dotty-example-project/src/main/scala-0.1", "/test/dotty-example-project/src/main/scala", "/test/dotty-example-project/src/main/java", "/test/dotty-example-project/target/scala-0.1/src_managed/main" ],
  "dependencyClasspath" : [ "/.ivy2/cache/ch.epfl.lamp/scala-library/jars/scala-library-0.1.2-RC1.jar", "/.ivy2/cache/ch.epfl.lamp/dotty-library_0.1/jars/dotty-library_0.1-0.1.2-RC1.jar", "/.ivy2/cache/org.scala-lang/scala-library/jars/scala-library-2.11.11.jar" ],
  "classDirectory" : "/test/dotty-example-project/target/scala-0.1/classes"
} ]

```

CallStack:
```
Starting server
Jun 02, 2017 9:18:41 AM org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint notify
INFO: Unsupported notification method: $/setTraceNotification
Jun 02, 2017 9:18:42 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification

java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:53)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:126)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:165)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:136)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:149)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:77)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51)
	... 11 more
Caused by: java.lang.NullPointerException
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:219)
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:211)
	at scala.collection.mutable.HashMap$$anon$3.next(HashMap.scala:146)
	at scala.collection.IterableLike$class.head(IterableLike.scala:107)
	at scala.collection.AbstractIterable.head(Iterable.scala:54)
	at dotty.tools.languageserver.DottyLanguageServer.driverFor(DottyLanguageServer.scala:83)
	at dotty.tools.languageserver.DottyLanguageServer.didOpen(DottyLanguageServer.scala:143)
	... 16 more

java.lang.NullPointerException
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:219)
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:211)
	at scala.collection.mutable.HashMap$$anon$3.next(HashMap.scala:146)
	at scala.collection.IterableLike$class.head(IterableLike.scala:107)
	at scala.collection.AbstractIterable.head(Iterable.scala:54)
	at dotty.tools.languageserver.DottyLanguageServer.driverFor(DottyLanguageServer.scala:83)
	at dotty.tools.languageserver.DottyLanguageServer.hover$$anonfun$1(DottyLanguageServer.scala:286)
	at dotty.tools.languageserver.DottyLanguageServer.liftedTree1$1(DottyLanguageServer.scala:107)
	at dotty.tools.languageserver.DottyLanguageServer.computeAsync$$anonfun$1(DottyLanguageServer.scala:112)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Jun 02, 2017 9:18:42 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint lambda$static$0
SEVERE: Internal error: java.lang.NullPointerException
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.NullPointerException
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:219)
	at scala.collection.mutable.HashTable$$anon$1.next(HashTable.scala:211)
	at scala.collection.mutable.HashMap$$anon$3.next(HashMap.scala:146)
	at scala.collection.IterableLike$class.head(IterableLike.scala:107)
	at scala.collection.AbstractIterable.head(Iterable.scala:54)
	at dotty.tools.languageserver.DottyLanguageServer.driverFor(DottyLanguageServer.scala:83)
	at dotty.tools.languageserver.DottyLanguageServer.hover$$anonfun$1(DottyLanguageServer.scala:286)
	at dotty.tools.languageserver.DottyLanguageServer.liftedTree1$1(DottyLanguageServer.scala:107)
	at dotty.tools.languageserver.DottyLanguageServer.computeAsync$$anonfun$1(DottyLanguageServer.scala:112)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 6 more

[Error - 9:18:42 AM] Request textDocument/hover failed.
  Message: Internal error, please look at the server's logs.
  Code: -32603 
```

smarter commented:
@HerringtonDarkholme My guess is that you have some scala files directly in the directory `/test/dotty-example-project`. If so, try moving them inside `/test/dotty-example-project/src/main/scala` (and I'll make sure to fix the IDE support so that this is not needed soon). Otherwise, could you post the output of `ls -R /test/dotty-example-project`, and the path of the file you tried to open ?

HerringtonDarkholme commented:
This is my `tree -L 2` output.

```
.
├── LICENSE
├── README.md
├── build.sbt
├── project
│   ├── build.properties
│   ├── plugins.sbt
│   ├── project
│   └── target
├── src
│   └── main
└── target
    ├── resolution-cache
    ├── scala-0.1
    └── streams
```

wxdong5211 commented:
![dottynpes](https://cloud.githubusercontent.com/assets/2654766/26709384/f2e9bed2-4784-11e7-8962-14b85b00ccbf.png)
me too on windoge 8

HerringtonDarkholme commented:
I have identified one problem.  A `for` comprehension will catch all Exception. https://github.com/lampepfl/dotty/blob/master/language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala#L65-L69

And alas, `InteractiveDriver` will initialize `Context` and thus `ScalaSettings` where `sys.props("user.dir")` will be queried.

However, if user has already opened `vscode`, `launchIDE` will reuse vscode instance instead of creating new instance in working directory. I personally open vscode from dock where `user.dir` is a `/`. Thus https://github.com/lampepfl/dotty/blob/352c09e481cb4893812c5f49959e587842448fe0/compiler/src/dotty/tools/dotc/config/ScalaSettings.scala#L124 throws and swallowed by `for`.

smarter commented:
@HerringtonDarkholme Thanks for the diagnostic! Can you try forcing code to run in a new instance using `code -n /test/dotty-example-project` and see if that helps? I'll make a PR to fix the exception when user.dir is `/`.

ouchxp commented:
@smarter `code -n /test/dotty-example-project` works for me.

HerringtonDarkholme commented:
Yes, fresh launching makes everything work.

smarter commented:
@wxdong5211 @HerringtonDarkholme Your issues should now be fixed if you use the latest nightly build of dotty (`scalaVersion := "0.2.0-bin-20170620-1d05796-NIGHTLY"` in your build) and the latest version of sbt-dotty (`addSbtPlugin("ch.epfl.lamp" % "sbt-dotty" % "0.1.3")` in `project/plugins.sbt`). Let me know if something is still broken!

HerringtonDarkholme commented:
It works fine for me! Thank @smarter !

wxdong5211 commented:
windoge ok too , @smarter 👍
